### PR TITLE
[bbs] fix incorrect revision fetch EXP-496

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

**TODO**
- [x] revert unit test changes

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 480df24</samp>

This pull request enhances the Bitbucket Server context parser by adding support for branches and tags in URLs, refactoring the tests and the API methods, and adding a new test script.

</details>

## Related Issue(s)

Fixes EXP-496

## How to test

- Add https://bitbucket.gitpod-dev.com as your git provider or join this ORG https://hw-bbs-revision.preview.gitpod-dev.com/orgs/join?inviteId=162b46e2-6ba7-46bd-a269-a84eece15076
- Create a workspace base [develop](https://bitbucket.gitpod-dev.com/users/huiwen/repos/mustard/browse?at=refs%2Fheads%2Fdevelop) branch
- Create a workspace base on [main](https://bitbucket.gitpod-dev.com/users/huiwen/repos/mustard/browse) branch
- They should both work

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-bbs-revision</li>
	<li><b>🔗 URL</b> - <a href="https://hw-bbs-revision.preview.gitpod-dev.com/workspaces" target="_blank">hw-bbs-revision.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-bbs-revision-gha.15783</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Review by filiptronicek — APPROVED** [review]
Works as advertised - both workspace descriptions as well as the branches they are checked out on 
image

File: components/server/src/bitbucket-server/bitbucket-server-api.ts [hunk 1]
```diff
@@ -190,6 +191,50 @@ export class BitbucketServerApi {
...
+        params: {
+            repoKind: "projects" | "users" | string;
+            owner: string;
+            repositorySlug: string;
+            branch: string;
+        },
+    ): Promise<BitbucketServer.Branch | undefined> {
+        // @see https://developer.atlassian.com/server/bitbucket/rest/v811/api-group-repository/#api-api-latest-projects-projectkey-repos-repositoryslug-branches-get
+        // @see https://bitbucket.gitpod-dev.com/rest/api/1.0/users/huiwen/repos/mustard/branches?filterText=develop
+        const queryParam = qs.stringify({
```
Comment (filiptronicek) [inline components/server/src/bitbucket-server/bitbucket-server-api.ts:205]: Is it expected that for invalid (not existing) branches the WS gets created on the default one? For GitHub it finds invalid branches
image

File: components/server/src/bitbucket-server/bitbucket-server-api.ts [hunk 2]
```diff
@@ -190,6 +191,50 @@ export class BitbucketServerApi {
...
+            filterText: params.branch,
```
Comment (filiptronicek) [inline components/server/src/bitbucket-server/bitbucket-server-api.ts:206]: Not sure if it's `true` by default or not, but we could include `boostMatches` to both improve performance and fix edge cases with having many branches which kindof match pushing the exact match to a further page in the pagination

Comment (filiptronicek) [reply components/server/src/bitbucket-server/bitbucket-server-api.ts:206]: https://developer.atlassian.com/server/bitbucket/rest/v811/api-group-repository/#api-api-latest-projects-projectkey-repos-repositoryslug-branches-get-request-Query%20parameters

File: components/server/src/bitbucket-server/bitbucket-server-api.ts (hunk 1 above)
Comment (filiptronicek) [reply components/server/src/bitbucket-server/bitbucket-server-api.ts:205]: Congrats on the number 500 btw 😆

File: components/server/src/bitbucket-server/bitbucket-server-api.ts (hunk 2 above)
Comment (mustard-mh) [reply components/server/src/bitbucket-server/bitbucket-server-api.ts:206]: Addressed 🙏

**Pull request merged**